### PR TITLE
chore: command actor refactor & add intg test [DET-9660]

### PR DESCRIPTION
**Pull request by carolinaecalderon**
## Description
Refactor the command module, and remove the actor system.

## Test Plan
See attached `command_intg_test.go`

## Commentary (optional)
I edited the `postgres_tasks.go` and `postgres_jobs.go` files in order to make their respective 'Add' tx functions public because I wanted to run `registerJobAndTask` as a tx in the command module. This way, if there's some failure to write to the db in any of the 3 operations in that method, the entire transaction is rolled back versus leaving some zombie entries in the db.

I also removed the vestigial `api.go`, `summary.go` (which was redundant with the command snapshot struct) & support for the NTSC handlers.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
DET-9660

### CODE REVIEW

File: master/internal/api_notebook.go
Comment (erikwilson) [inline master/internal/api_notebook.go:56]: should this be "notebooks"? it might be good to keep those constants around

Comment (erikwilson) [reply master/internal/api_notebook.go:56]: rather, it looks like there is an ask or actor in this that might need to be updated to remove notebooksAddr

Comment (carolinaecalderon) [reply master/internal/api_notebook.go:56]: oh this is just vestigial code that I edited to make the code compile partway through -- it's not necessary now, so I'll delete that line

File: master/internal/command/api.go
Comment (stoksc) [inline master/internal/command/api.go:34]: if you need a hand with this just let me know

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:114]: i'd make a `newCommandService` and have it do this, or just make `restoreAllCommands` do the locking.

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:300]: you're taking a tx and not using it

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:332]: this looks right to me

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:386]: i'd have a different strategy for stuff like this, maybe we (you me erik) can talk offline now that I've looked?

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:385]: i think we can unregister the job immediately, not in gc

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:425]: nit https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:476]: better signature?
```suggestion
func (c *command) deleteIfInWorkspace(workspaceID int64) {
```

File: master/internal/command/command_manager.go
Comment (stoksc) [inline master/internal/command/command_manager.go:24]: getAll -> list

File: master/internal/command/command_manager.go
Comment (stoksc) [inline master/internal/command/command_manager.go:39]: return the error dont log here and in a lot of places throughout this file (we're calling .toCommand on nil or zero value).

File: master/internal/command/notebook_manager.go
Comment (stoksc) [inline master/internal/command/notebook_manager.go:28]: can we possibly push this users/userIds agg + filtering by it into `getAllNTSCType` to reduce duplication?

File: master/internal/command/notebook_manager.go
Comment (stoksc) [inline master/internal/command/notebook_manager.go:41]: more return vs log error stuff (last comment i'll leave about it, just pointing out it is in multiple files)

**Review by stoksc — COMMENTED** [review]
mostly looks good, saw a few bugs and a few things we should chat about over teams though.

File: master/internal/command/command.go
Comment (carolinaecalderon) [reply master/internal/command/command.go:300]: fixing

File: master/internal/command/command.go
Comment (carolinaecalderon) [reply master/internal/command/command.go:114]: Since I think we want locks in public functions, then I think we _shouldn't_ lock in `restoreAllCommands`.

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:383]: can you put command service into a separate file from command?

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:107]: can we get rid of this and "init" it in `new`? i don't like having a `new` that construct an incorrect/incomplete thing. it is an unintuitive/easily misused API.

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:144]: use `fmt.Errorf` for all new code

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:242]: did we change the order of this code (this runs after the `c.restored` block) for any reason? if not, put it back to minimize risk?

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:118]: seems like we should return here, not iterate commands and try to start them still

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:323]: i liked the old function `c.persistAndEvictContextDirectoryFromMemory()`. made it very clear by name why this last, subtle line was so important. can we bring it back and just pass it a `tx`?

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:123]: let's just make `commandFromSnapshot` call `startCmd` like `createGeneric..` does. more of "things should be constructed correctly by their constructor, not have methods users need to call to make them correct".

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:219]: needs locks

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:394]: i don't see this method as very helpful, as opposed to users just calling `createGenericCommand` directly

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:422]: nit: indent error flow
```suggestion
	if c.taskType != cmdType {
		cs.syslog.Debugf("getNTSC: type mismatch: %s/%s", cmdType, c.taskType)
		return nil, fmt.Errorf("get NTSC %s not found", cmdID)
	}
	return c, nil
	
```

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:426]:
```suggestion
func (cs *commandService) listByType(
```
personally, i dislike "NTSC".

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:441]: i think in a lot of places we need to be locking this map?

File: master/internal/command/command_manager.go
Comment (stoksc) [inline master/internal/command/command_manager.go:96]: we need at least a read lock over command state here right?

**Review by stoksc — COMMENTED** [review]
it may make sense to pair review/iterate on this, if you are up for it, mostly to make sure we get the locking right.

File: master/internal/command/command.go
Comment (carolinaecalderon) [reply master/internal/command/command.go:107]: The issue is that I can't figure out how to initialize/define the command service's db/rm fields _outside_ of calling an Init function in core.go. We need the command service to store these values because each command it creates/tracks needs a db & rm to start its allocation -- and we can't easily find this value when we ask to launch the NTSC.
I'll think some more. I think what you did with allocation service was more simple since you didn't need to pass variables to define the singleton service.

Comment (stoksc) [reply master/internal/command/command.go:107]: can you just make a `NewCommandService(db *db.PgDB, rm rm.ResourceManager) *CommandService` and then `SetCommandService(*CommandService)` to init the singleton? i don't love the pattern i did with allocation, i'll change it.

File: master/internal/command/command.go
Comment (carolinaecalderon) [reply master/internal/command/command.go:219]: reworked it, but `createGenericCommand` is a private function, and wherever it's called (the public 'LaunchNotebook/shell/command/tb' functions) locks the command service before calling it.

File: master/internal/command/command.go
Comment (carolinaecalderon) [reply master/internal/command/command.go:441]: `listByType` (prev `listNTSCType`) is called by the public GetNotebook/Command/Shell/Tb functions -- and in those public functions I locked the command service.

File: master/internal/command/command_manager.go
Comment (carolinaecalderon) [reply master/internal/command/command_manager.go:96]: I don't remember why I took the lock off -- I think it was also the whole 'private vs public' function thing. maybe I'm following that rule too strictly.

File: master/internal/command/command_manager.go
Comment (stoksc) [inline master/internal/command/command_manager.go:63]:
```suggestion
// KillCommand sends a kill signal to the command's allocation.
```

File: master/internal/command/command_manager.go
Comment (stoksc) [inline master/internal/command/command_manager.go:100]: move this method to `command.go`, all the others to `command_service.go` and remove this file?

File: master/internal/command/command_manager.go
Comment (stoksc) [inline master/internal/command/command_manager.go:47]:
```suggestion
// GetCommand looks up a command by ID returns a summary of the its state and configuration.
```
these comments could be a little more descriptive.

File: master/internal/command/command_service.go
Comment (stoksc) [inline master/internal/command/command_service.go:38]: i'd prefer to export `commandService` and give it a proper constructor but i wouldn't block on it.

**Review by stoksc — APPROVED** [review]
lgtm

**Comment by stoksc** [comment]
looks good overall, though obvious make sure the tests pass and you are confident their aren't breakages. i would run some select longrunning tests before merging this one.

File: master/internal/command/command_manager.go
Comment (carolinaecalderon) [reply master/internal/command/command_manager.go:100]: disagree, because the methods in each XYZ_manager.go file take in API requests & return API responses. To me, mixing up the command_service methods with API-related methods doesn't make as much sense as separating it out.

Comment (stoksc) [reply master/internal/command/command_manager.go:100]: yep i agree, i just mixed up `commandService` and `CommandService`.. kinda confusing names

Comment (stoksc) [reply master/internal/command/command_manager.go:100]: oh well, i do this this should be in `command.go` though. why would a method on command not be in the file the struct is in? it doesn't have to but it is confusing to me.

File: master/internal/command/command_service.go
Comment (stoksc) [inline master/internal/command/command_service.go]: i think this should return not found, too

File: master/internal/command/command.go
Comment (erikwilson) [inline master/internal/command/command.go:173]: Should we return this error? And change to persist before StartAllocation? cc @stoksc

File: master/internal/command/command.go
Comment (erikwilson) [inline master/internal/command/command.go:47]: nit: since we are locking these functions it might be best to capitalize them as it signifies they are being called externally.

File: master/internal/command/command.go
Comment (carolinaecalderon) [reply master/internal/command/command.go:173]: +1. can change either way

File: master/internal/command/command.go
Comment (carolinaecalderon) [reply master/internal/command/command.go]: capitalized!

File: master/internal/command/command_intg_test.go
Comment (stoksc) [inline master/internal/command/command_intg_test.go:176]: this is in enough places we could stand to have a dedicated "mockRMThatAlwaysJustWorks" helper.

File: master/internal/command/command_service.go
Comment (stoksc) [inline master/internal/command/command_service.go:22]: I'd prefer all the methods for this struct in the same file. I was trying to vet the locking and it is hard without this.

Also, i think there are some locking optimizations we can do, but let's in a follow up.

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:264]: we need this, right? else we are susceptible to a lock cycle
```suggestion
	go jobservice.DefaultService.UnregisterJob(c.jobID)
```

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:267]: i think we can rm this. looks like you moved it to the allocation exited callback where it always should've been, from GC.

File: master/internal/command/command.go
Comment (stoksc) [inline master/internal/command/command.go:296]: style nit, makes a lot more sense to have the caller call ToV1Command().WorkspaceID and KillNTSC, and for them to impl this logic client side than adding this very specific API to Command that it is unlikely for someone else to have a use for.

File: master/internal/core.go
Comment (stoksc) [inline master/internal/core.go:1088]: I would usually call this init. I would expect Set to take a CommandService, not construct one for the caller.

File: master/internal/core.go
Comment (stoksc) [inline master/internal/core.go:1090]: it is a style thing, but i'd rather have a NewCommandService and constructs the command service and makes sure it is restored before we call SetDefaultCmdService to it is compiler verified no one can call the default command service before it is correctly instantiated

**Review by stoksc — APPROVED** [review]
lgtm. I only found one blocking bug, the use of `go` for callback comment.

File: master/internal/command/command.go
Comment (carolinaecalderon) [reply master/internal/command/command.go:264]: yeah that should have the `go`, not sure at what point I removed that, but it should be there

File: master/internal/command/command_intg_test.go
Comment (carolinaecalderon) [reply master/internal/command/command_intg_test.go:176]: worth another ticket, but imo not here

Comment (carolinaecalderon) [reply master/internal/command/command_intg_test.go:176]: https://hpe-aiatscale.atlassian.net/browse/DET-9967

**Pull request merged**
